tools/serve/serve.py

```diff
@@ -204,6 +207,65 @@ def _script_replacement(self, key, value):
         return None
 
 
+class HtmlScriptInjectorHandlerWrapper:
+    def __init__(self, inject="", wrap=None):
+        self.inject = inject
+        self.wrap = wrap
+
+    def __call__(self, request, response):
+        self.wrap(request, response)
+        # If the response content type isn't html, don't modify it.
+        supported_mime_types = {b"text/html", b"application/xhtml+xml"}
+        if not isinstance(response.headers, ResponseHeaders) or not response.headers.get("Content-Type")[0] in supported_mime_types:
+            return response
+
+        # Skip injection on custom streaming responses.
+        if not isinstance(response.content, (bytes, str, IOBase)) and not hasattr(response, "read"):
+            return response
+
+        # Otherwise, inject the script after the document's doctype.
+        data = "".join(item.decode(response.encoding) for item in response.iter_content(read_file=True))
+
+        # Tokenize and find the position of the first content (e.g. after the
+        # doctype if one is present).
+        token_types = html5parser.tokenTypes
+        after_tags = {"html", "head"}
+        before_tokens = {token_types["EndTag"], token_types["EmptyTag"],
+                         token_types["Characters"]}
+        error_tokens = {token_types["ParseError"]}
+
+        tokenizer = html5parser._tokenizer.HTMLTokenizer(data)
+        stream = tokenizer.stream
+        offset = 0
+        error = False
+        for item in tokenizer:
+            if item["type"] == token_types["StartTag"]:
+                if not item["name"].lower() in after_tags:
+                    break
+            elif item["type"] in before_tokens:
```

Comment:
Since `token_types["Characters"]` is in `before_tokens` and we break below, does this mean that `offset` will never go beyond the first text node, even whitespace? Or is `token_types["Characters"]` not for text nodes?

Reply:
Whitespace is skipped, but text nodes will not be. I imagine it would be fine in general to skip past character tokens. The main thing we need the script to occur before is any other script.

Reply:
Oh, so `token_types["Characters"]` is for whitespace inside of tags and things like that? That makes sense then.

Reply:
`token_types["SpaceCharacters"]` is a different thing: https://github.com/html5lib/html5lib-python/blob/master/html5lib/constants.py#L2921-L2930